src/language-js/printer-estree.js

```diff
@@ -1807,8 +1807,24 @@ function printPathNoParens(path, options, print, args) {
       if (n.value) {
         let res;
         if (isStringLiteral(n.value)) {
-          const value = rawText(n.value);
-          res = '"' + value.slice(1, -1).replace(/"/g, "&quot;") + '"';
+          const innerValue = rawText(n.value).slice(1, -1);
+          const hasSingle = innerValue.includes("'");
+          const hasDouble = innerValue.includes('"');
+
+          // If we need to escape quotes, just use the JS handler
+          if (hasSingle && hasDouble) {
+            res = concat(["{", path.call(print, "value"), "}"]);
```

Comment:
Unfortunately, this isn’t correct either. `<e attr="&nbsp;" />` is not equivalent to `<e attr={"&nbsp;"} />`. And I don’t think we should try converting HTML escapes into JS escapes. Again, I don’t think it is possible to get into this situation when the AST comes from a parser.

Reply:
ok I'm not experienced with all the parsing quirks and [I'm getting conflicting replies](https://github.com/prettier/prettier/pull/4798#discussion_r199989155) on how to handle this.

From your comments it seems that it's not feasible to use the `printString` method itself to do the work, but we want to get the same functionality as `printString` which lease the option of abstracting the preferred code selection logic 

[I opened a 2nd branch with that change](https://github.com/smirea/prettier/compare/feat-option-singleQuote-jsx...smirea:feat-option-singleQuote-jsx-escape?expand=1) can I get your thoughts on it before updating this one? I'm trying to get a grasp on this before spamming everyone else with more updates

Reply:
Looks great! The only change I would make is to remove unnecessary escapes (i.e. change `'&apos;'` to `"'"` instead of `"&apos;"`), but I’m not at a computer to check if your branch does that already.

Reply:
That second branch looks good.

If you’re interested, here’s some stuff to read on the quirkyness of escaping in JSX:

https://github.com/yannickcr/eslint-plugin-react/issues/1449
https://github.com/yannickcr/eslint-plugin-react/pull/1458

Reply:
@j-f1 it did not, it does now
@lydell thanks for the feadback + quirkiness docs, it was an interesting read!